README.md

```diff
@@ -1,6 +1,463 @@
+      "name": "Earnings Transcripts",  
+      "description": "Contains earnings transcripts for a ticker",
+      "data": {
+        "content": "<data>"  # <-- the data of the widget
+      },
+      "metadata": {
+        "symbol": "AAPL",
+        "period": "quarter",
+        "source": "Intrinio",
+        "lastUpdated": 1728998071322
+      }
+    },
+    ...
+  ],
+  ...
+}
+```
+
+#### `widgets`
+
+This is an array of widgets that are currently visible on the active dashboard
+on Terminal Pro.
+**This is only useful if you're planning on implementing function calling in
+*your custom copilot** (which is recommended, but not required), which
+allows it to request widget data from the user's currently-active dashboard on
+OpenBB Terminal.
+
+```python
+{
+  ...
+  "widgets": [
+    {
+      "uuid": "c276369e-e469-4689-b5fe-3f8c76f7c45a",
+      "name": "Stock Price Quote Widget",
+      "description": "Contains the current stock price of a ticker",
+      "metadata": {
+        "ticker": "AAPL"
+      }
+    },
+    {
+      "uuid": "9f8e7d6c-5b4a-3c2e-1d0f-9e8d7c6b5a4b",
+      "name": "Financial Ratios Widget",
+      "description": "Displays key financial ratios for a company",
+      "metadata": {
+        "ticker": "AAPL",
+        "period": "TTM"
+      }
+    },
+    ...
+  ],
+  ...
+}
+```
+
+## Response Schema
+
+Your custom copilot must respond to OpenBB Terminal's request using Server-Sent Events (SSEs).
+
+OpenBB Terminal can process the following SSEs:
+
+- `copilotMessageChunk`: Used to return streamed copilot tokens (partial responses) back to OpenBB Terminal. These responses can be streamed as they are generated.
+- `copilotFunctionCall`: Used to request data (e.g., widget data) or perform a specific function. This instructs Terminal Pro to take further action on the client's side. This is only necessary if you're planning on implementing function calling in your custom copilot.
+
+#### `copilotMessageChunk`
+The message chunk SSE has the following format:
+
+```
+event: copilotMessageChunk
+data: {"delta":"H"}  # <-- the `data` field must be a JSON object.
+```
+The `delta` must be a string, but can be of any length. We suggest streaming
+back each chunk you receive from your LLM as soon as it's generated as a `delta`.
+
+For example, if you wanted to stream back the message "Hi", you would send the
+following SSEs:
+
+```
+event: copilotMessageChunk
+data: {"delta":"H"}
+
+event: copilotMessageChunk
+data: {"delta":"i"}
+
+event: copilotMessageChunk
+data: {"delta":"!"}
+```
+
+#### `copilotFunctionCall` (only required for function calling)
+The function call SSE has the following format:
+
+```
+event: copilotFunctionCall
+data: {"function":"get_widget_data","input_arguments":{"widget_uuid":"c276369e-e469-4689-b5fe-3f8c76f7c45a"}}
+```
+
+Again, the `data` field must be a JSON object. The `function` field is the name
+of the function to be called (currently only `get_widget_data` is supported),
```

Comment:
What about `get_direct_retrieval_data`?

Reply:
This is currently not supported on custom copilots (but it's something we can add -- although I'm not sure how many users would want this, since they'll already have access to the data sources directly on their machine, and so could make a request to the data directly). 

We'll need to modify the way OpenBB Copilot works first to make sure things are compatible. For example, we'll need to fetch the built-in `widgets.json` dynamically on start-up, so that custom copilots can follow the same behaviour.